Vladislav Golovkin/0/main.rb

```diff
@@ -0,0 +1,180 @@
+require 'roo'
+require 'spreadsheet'
+
+# product properties
+class ProductData
+  attr_reader :date, :price
+  attr_writer :date, :price
+
+  def initialize
+    @price = 0
+    @date = Date.new
+  end
+end
+
+# collection of products
+class Products
+  def initialize
+    @dictionary = {}
+  end
+
+  def add_entry(product_name, date, price)
+    new_entry = ProductData.new
+    new_entry.date = date
+    new_entry.price = price
+
+    new_entry.price /= 10_000 if date.year < 2017
+
+    product_name = product_name.to_s.squeeze(' ')
+    reg_exp = /^[\dA-Z_]+$/
+    product_name = product_name.gsub(reg_exp, '')
+
+    @dictionary[product_name] = [] unless @dictionary.key?(product_name)
+    @dictionary[product_name] << new_entry
+  end
+
+  def get_same_price_products(latest_product)
+    result = []
+    @dictionary.each do |name, products|
+      products.each do |product|
+        price = latest_product.price
+        next unless product.date == Date.new(2018, 10, 1) &&
+                    product.price.between?(price - 0.2, price + 0.2)
+
+        result << "#{name.capitalize} with price #{product.price} BYN"
+      end
+    end
+    result
+  end
+
+  def process_request(input)
+    @dictionary.each do |key, value|
```

Comment:
|key, value| -- what is key? what is value?

maybe -- |product_name, products|?